kahuna/public/js/search-query/service.js

```diff
@@ -0,0 +1,32 @@
+import angular from 'angular';
+import Rx from 'rx';
+import '../util/rx';
+
+export const searchQueryService = angular.module('gr.searchQuery.service', ['util.rx']);
+
+searchQueryService.factory('searchQueryService', [function() {
+    const query$ = new Rx.Subject();
+
+    // TODO: When we update rx, we need to swap the seed / acc around
+    const q$ = query$.scan('', (state, fn) => fn(state).trim());
+
+    function hasLabel(q, label) {
+        // TODO: It'd be nice to build this up from an API
+        const r = new RegExp(`[#|label\:]${label}\\b`);
+        return r.test(q);
+    }
+
+    function addLabel(label) {
+        query$.onNext(q => hasLabel(q, label) ? q : `${q} #${label}`);
```

Comment:
@tsop14 was working on a fix for labels that contain spaces. I can't remember if it was released or not?